milabench/executors.py

```diff
@@ -0,0 +1,612 @@
+
+        Arguments:
+            **kwargs: some `Executor` might need an argument to dynamically
+                      generate the list of command line's arguments
+        """
+        if self.exec:
+            return self._argv(**kwargs) + self.exec.argv(**kwargs)
+        return self._argv(**kwargs)
+
+    def kwargs(self) -> Dict:
+        """Return the `Executor`'s kwargs to send to `BasePackage.execute()`
+        merged with the embeded `Executor`, if any
+
+        The `Executor`'s kwargs will take priority over the embeded `Executor`'s
+        kwargs
+        """
+        kwargs = self._kwargs
+        if self.exec:
+            kwargs = {**self.exec.kwargs(), **kwargs}
+        return kwargs
+
+    def commands(self) -> Generator[Tuple[pack.BasePackage, List, Dict], None, None]:
+        """Return a tuple of the leaf's `BasePackage`, the `Executor`'s list of
+        command line's arguments and the `Executor`'s kwargs to send to
+        `BasePackage.execute()`
+        """
+        yield self.pack, self.argv(), self.kwargs()
+
+    async def execute(self, timeout=False, timeout_delay=600, **kwargs):
+        """Execute all the commands and return the aggregated results
+        """
+        coro = []
+
+        for pack, argv, _kwargs in self.commands():
+            await pack.send(event="config", data=pack.config)
+            await pack.send(event="meta", data=machine_metadata())
+
+            if hasattr(pack, 'override_run'):
+                # TODO: This is only to hack the CI and have the test passing
+                # using the old system. Remove override_run() and use `Executor`
+                # instead 
+                fut = pack.override_run()
+            else:
+                pack.phase = "run"
+                fut = pack.execute(*argv, **{**_kwargs, **kwargs})
+
+            coro.append(fut)
+
+            if timeout:
+                delay = pack.config.get("max_duration", timeout_delay)
+                asyncio.create_task(force_terminate(pack, delay))
+
+        return await asyncio.gather(*coro)
+
+    def _argv(self, **kwargs) -> List:
+        del kwargs
+        return []
+
+
+# Leafs
+class CmdExecutor(Executor):
+    """Execute a command
+
+    Arguments:
+        pack: `BasePackage`'s instance from which `execute()` will be used to
+              perform the command
+        *cmd_argv: command line arguments list to execute
+        **kwargs: kwargs to be passed to the `pack.execute()`
+    """
+    def __init__(
+            self,
+            pack:pack.BasePackage,
+            *cmd_argv,
+            **kwargs
+    ) -> None:
+        if isinstance(pack, Executor):
+            raise ValueError(f"{self.__class__.__name__} does not accept nested"
+                             f" {Executor.__class__.__name__}")
+        super().__init__(
+            pack,
+            **kwargs
+        )
+        self.cmd_argv = cmd_argv
+
+    def _argv(self, **kwargs) -> List:
+        del kwargs
```

Comment:
?

Reply:
useless, only to remove some warnings about unused variables. But I can del the `del` line

Reply:
Normally you shouldn't get warnings about unused parameters, I don't think. Is this flake8 or just the IDE? You can try using `**_`, which linters usually acknowledge as a convention for "I'm not using this".